app-engine/java/testSrc/com/google/cloud/tools/intellij/appengine/java/cloud/AppEngineArtifactDeploymentSourceProviderTest.java

```diff
@@ -0,0 +1,183 @@
+import com.intellij.packaging.artifacts.ArtifactManager;
+import com.intellij.packaging.artifacts.ModifiableArtifact;
+import com.intellij.packaging.artifacts.ModifiableArtifactModel;
+import com.intellij.packaging.elements.PackagingElement;
+import com.intellij.packaging.elements.PackagingElementFactory;
+import com.intellij.packaging.impl.artifacts.ArtifactManagerImpl;
+import com.intellij.packaging.impl.artifacts.PlainArtifactType;
+import com.intellij.packaging.impl.elements.ArchivePackagingElement;
+import com.intellij.remoteServer.configuration.deployment.DeploymentSource;
+import com.intellij.testFramework.fixtures.IdeaProjectTestFixture;
+import java.util.List;
+import java.util.Optional;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.mockito.Mock;
+
+/** Tests for {@link AppEngineArtifactDeploymentSourceProvider} */
+public class AppEngineArtifactDeploymentSourceProviderTest {
+  @Rule public final CloudToolsRule cloudToolsRule = new CloudToolsRule(this);
+  @TestFixture private IdeaProjectTestFixture testFixture;
+
+  @TestModule private Module module1;
+  @TestModule private Module module2;
+
+  @TestService @Mock AppEngineProjectService projectService;
+
+  private AppEngineArtifactDeploymentSourceProvider artifactDeploymentSourceProvider;
+
+  @Before
+  public void setUp() {
+    artifactDeploymentSourceProvider = new AppEngineArtifactDeploymentSourceProvider();
+  }
+
+  @Test
+  public void getDeploymentSources_withOnlyStandardArtifacts_returnsAllSources() {
+    makeStandard(module1);
+    makeStandard(module2);
+
+    assertDeploymentSourcesPresent(AppEngineEnvironment.APP_ENGINE_STANDARD);
+  }
+
+  @Test
+  public void getDeploymentSources_withOnlyFlexibleArtifacts_returnsAllSources() {
+    makeFlexible(module1);
+    makeFlexible(module2);
+
+    assertDeploymentSourcesPresent(AppEngineEnvironment.APP_ENGINE_FLEX);
+  }
+
+  @Test
+  public void getDeploymentSources_withStandardAndFlexibleArtifacts_returnsAllSources() {
+    makeStandard(module1);
+    makeFlexible(module2);
+
+    List<DeploymentSource> deploymentSources =
+        artifactDeploymentSourceProvider.getDeploymentSources(testFixture.getProject());
+
+    assertThat(deploymentSources.size()).isEqualTo(2);
+
+    DeploymentSource deploymentSource1 = deploymentSources.get(0);
+    DeploymentSource deploymentSource2 = deploymentSources.get(1);
+
+    assertThat(deploymentSource1 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(deploymentSource2 instanceof AppEngineArtifactDeploymentSource).isTrue();
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource1).getEnvironment())
+        .isEqualTo(AppEngineEnvironment.APP_ENGINE_STANDARD);
+    assertThat(((AppEngineArtifactDeploymentSource) deploymentSource2).getEnvironment())
+        .isEqualTo(AppEngineEnvironment.APP_ENGINE_FLEX);
+  }
+
+  private void assertDeploymentSourcesPresent(AppEngineEnvironment environment) {
```

Comment:
Could we expand this method a bit to avoid almost identical code in lines 91:104?

Reply:
done